common/src/stack/command/stack/commands/load/json/plugin_software.py

```diff
@@ -0,0 +1,158 @@
+		# add boxes
+		if import_data['box']:
+			for box in import_data['box']:
+				parameters = [
+					f'{box["name"]}',
+					f'os={box["os"]}',
+				]
+				self.owner.try_command('add.box', parameters, f'adding box {box["name"]}', 'exists')
+
+		# add pallets
+		if import_data['pallet']:
+			# create a list of pallets to add, excluding ones already present on the system
+			pallets_to_add = []
+			existing_pallet_data = json.loads(self.owner.command('list.pallet', [ 'output-format=json', 'expanded=true' ]))
+
+
+			# iterate through each pallet, and only add it if not aleady on the system
+			for pallet in import_data['pallet']:
+				for existing_pallet in existing_pallet_data:
+					if (pallet['name'] == existing_pallet['name'] and
+					    pallet['version'] == existing_pallet['version'] and
+					    pallet['release'] == existing_pallet['release'] and
+					    ' '.join(pallet['boxes']) == existing_pallet['boxes']):
+						self.owner.log.info(f'The pallet {pallet["name"]} already exists on the system. skipping.')
+						break
+				else:
+					pallets_to_add.append(pallet)
+
+
+			# now we can go ahead adding the pallets without wasting time downloading duplicate pallets
+			for pallet in pallets_to_add:
+				pallet_dir =  pallet['url']
+				if pallet_dir == None:
+					# if we have no url to fetch the pallet from we cannot add it
+					raise CommandError(self.owner, f'error adding pallet {pallet["name"]} {pallet["version"]}: no url found')
+
+					# the following code is now unreachable, does it have any value?
+					# now that it is impossible for there to be a pallet added to the database that down not have a url
+
+					# in the rare case that this pallet has no url but also currently exists in the database
+					# we need to check to see if it needs to be added to any boxes
+#					if not pallet['boxes']:
+#						continue
+#					pallet_data = self.owner.command('list.pallet', [ 'output-format=json', 'expanded=true' ])
+#					if pallet_data:
+#						pallet_data = json.loads(pallet_data)
+#						for item in pallet_data:
+#							if pallet['name'] == item['name']:
+#								# we have now deduced that the pallet both is currently in the database and that it has boxes that it needs to be added to
+#								for box in pallet['boxes']:
+#								parameters = [
+#									pallet['name'],
+#									f'release={pallet["release"]}',
+#									f'box={box}',
+#								]
+#								self.owner.try_command('enable.pallet', parameters, f'enabling {pallet} in {box}', 'exists')
+#						# we have finished with this pallet
+#						continue
+
+
+				parameters = [pallet_dir]
+				# if there are credentials, add them to parameters
+				if pallet['urlauthUser'] and pallet['urlauthPass']:
+					parameters.append(f'username={pallet["urlauthUser"]}')
+					parameters.append(f'password={pallet["urlauthPass"]}')
+
+
+				# we do not want to add the stacki pallet, but we want to put it in the necessary boxes
```

Comment:
Actually, I was just thinking about this.  My worry was overwriting the existing stacki pallet that the Frontend was installed with, but maybe that's not an issue after all.  We do need to add the stacki pallet if it's for a different version (for example, to serve SLES11 installs).  Let's do a quick re-think on this on Monday.